File: facebook_service/service.rb
require "rubygems"
require 'sinatra'
require 'json'
require 'bunny'
PORT = 8080
puts "Starting server on port #{PORT}"
connection = Bunny.new
connection.start

channel  = connection.create_channel
# topic exchange name can be any string
exchange = channel.topic("facebook_events", :durable => true)
channel.queue("location", :durable => true).bind(exchange, :routing_key => "location")

get '/event' do
  raw = request.body.read
  data = JSON.parse raw
  data["entry"].each do |entry| 
    entry["changed_fields"].each do |field|
      exchange.publish(raw, :routing_key => field)
    end
  end
  return '{"success":true}'
end

File: facebook_service/service.js
var amqp = require('amqp');
var http = require('http');
var cluster = require('cluster');

var connection = amqp.createConnection({url:"amqp://guest:guest@localhost:5672"});
connection.on('ready', function () {
  connection.queue('location', {durable: true}, function(q){
      q.bind('location');
  });
});

var app = function(request,res){
  var body = '';

  request.on('data', function (data) {
    body += data;
  });

  request.on('end', function () {
    var user = JSON.parse(body);
    connection.publish(user.entry[0].changed_fields[0], JSON.stringify(user));
    res.writeHeader(200, {"Content-Type": "application/json"});
    res.write('{"success":true}')
    res.end();
  });
};

if (cluster.isMaster) {
  // Fork workers.
  for (var i = 0; i < 4; i++) {
    cluster.fork();
  }
} else {
  console.log("Starting server on port 8080")
  http.createServer(app).listen(8080);
}
